ssl/quic/quic_record_rx.c

```diff
@@ -253,6 +275,193 @@ void ossl_qrx_inject_urxe(OSSL_QRX *qrx, QUIC_URXE *urxe)
+     * The packet header has been unprotected in place, so we can just reuse the
+     * PACKET buffer. The header ends where the payload begins.
+     */
+    aad_len = rxe->hdr.data - sop;
+
+    /* Ensure the RXE buffer size is adequate for our payload. */
+    if ((rxe = qrx_reserve_rxe(&qrx->rx_free, rxe, rxe->hdr.len + i)) == NULL)
+        goto malformed;
+
+    /*
+     * We decrypt the packet body to immediately after the token at the start of
+     * the RXE buffer (where present).
+     *
+     * Do the decryption from the PACKET (which points into URXE memory) to our
+     * RXE payload (single-copy decryption), then fixup the pointers in the
+     * header to point to our new buffer.
+     *
+     * If decryption fails this is considered a permanent error; we defer
+     * packets we don't yet have decryption keys for above, so if this fails,
+     * something has gone wrong with the handshake process or a packet has been
+     * corrupted.
+     */
+    dst = (unsigned char *)rxe_data(rxe) + i;
+    if (!qrx_decrypt_pkt_body(qrx, dst, rxe->hdr.data, rxe->hdr.len,
+                              &dec_len, sop, aad_len, rxe->pn, QUIC_ENC_LEVEL_INITIAL,
+                              rxe->hdr.key_phase, &rx_key_epoch))
+        goto malformed;
+
+    /*
+     * -----------------------------------------------------
+     *   IMPORTANT: ANYTHING ABOVE THIS LINE IS UNVERIFIED
+     *              AND MUST BE TIMING-CHANNEL SAFE.
+     * -----------------------------------------------------
+     *
+     * At this point, we have successfully authenticated the AEAD tag and no
+     * longer need to worry about exposing the PN, PN length or Key Phase bit in
+     * timing channels. Invoke any configured validation callback to allow for
+     * rejection of duplicate PNs.
+     */
+    if (!qrx_validate_hdr_late(qrx, rxe))
+        goto malformed;
+
+    pkt_mark(&urxe->processed, 0);
+
+    /*
+     * Update header to point to the decrypted buffer, which may be shorter
+     * due to AEAD tags, block padding, etc.
+     */
+    rxe->hdr.data       = dst;
+    rxe->hdr.len        = dec_len;
+    rxe->data_len       = dec_len;
+    rxe->datagram_len   = datagram_len;
+    rxe->key_epoch      = rx_key_epoch;
+
+    /* We processed the PN successfully, so update largest processed PN. */
+    pn_space = rxe_determine_pn_space(rxe);
+    if (rxe->pn > qrx->largest_pn[pn_space])
+        qrx->largest_pn[pn_space] = rxe->pn;
+
+    /* Copy across network addresses and RX time from URXE to RXE. */
+    rxe->peer           = urxe->peer;
+    rxe->local          = urxe->local;
+    rxe->time           = urxe->time;
+    rxe->datagram_id    = urxe->datagram_id;
+
+    /*
+     * Move RXE to pending. this effectively injects packet to ch_rx().
+     * The ch_rx() will be invoked with next tick on channel.
+     */
+    ossl_list_rxe_remove(&qrx->rx_free, rxe);
+    ossl_list_rxe_insert_tail(&qrx->rx_pending, rxe);
```

Comment:
can you just use ossl_qrx_inject_urxe() here, like we did in port_default_packet_handler()?  Doing so has the side effects of setting urxe->processed/hpr_removed/deferred and calling the msg_callback, which (maybe) you want to preserve?

Reply:
no I can't use this function here. This function is used by port to inject packet to QRX to insert the packet to qrx->urx_pending queue. The packet there sits there waiting for decryption. Howver to validate the packet the port decrypted the packet already, so we must move the packet to qrx->rx_pending queue. I think it deserves better comment. I will add one.